Introduce 'ClientSettings'; overhaul user system preferences

DanVanAtta opened the pull request:
## Change summary
* Rewrite the settings window, backed by an enum driven configuration.
* Redo the system preferences framework. Also now 'enum' backed
* Start killing off SwingComponents, add builders to replace those APIs

---------- 

## Also Notable
* Simplify the proxy code. Remove command line option to configure proxy, can now only be set from the UI. 
* Added a new "Show Beta Features" flag to the settings.
* Backup lobby mechanism updated, no more backup URL in props file. Instead we store the URL we last connected to, and use that if we cannot connect to github to get the lobby props file (which contains the lobby URL) 
-------------

Preview of new settings window UI:

![window_example](https://user-images.githubusercontent.com/12397753/28650277-82d5edca-722f-11e7-9ac9-b53e3cfc067f.png)

DanVanAtta commented on src/main/java/games/strategy/engine/framework/ArgParser.java:
```diff
@@ -30,14 +29,15 @@ public static boolean handleCommandLineArgs(
         return false;
       }
     }
+    ClientSetting.flush();
```
Needed because we call `setSystemProperty` previously. The `flush` is a slow call so it is outside of the for loop.

DanVanAtta commented on src/main/java/games/strategy/engine/framework/GameRunner.java:
```diff
@@ -122,8 +108,8 @@
       {TRIPLEA_GAME_PROPERTY, TRIPLEA_SERVER_PROPERTY, TRIPLEA_CLIENT_PROPERTY, TRIPLEA_HOST_PROPERTY,
           TRIPLEA_PORT_PROPERTY, TRIPLEA_NAME_PROPERTY, TRIPLEA_SERVER_PASSWORD_PROPERTY, TRIPLEA_STARTED,
           TRIPLEA_LOBBY_PORT_PROPERTY,
-          LOBBY_HOST, LOBBY_GAME_COMMENTS, LOBBY_GAME_HOSTED_BY, TRIPLEA_ENGINE_VERSION_BIN, HttpProxy.PROXY_HOST,
```
We do not run bots or lobby with a proxy arg, given it can be updated via UI, we really do not need command line args for it. Simplifying here by removing them. 

Proxy configuration was stored in the engine preferences window, to move this to the settings window required for me to go through the proxy code. There are some further updates/simplfications to the proxy code

DanVanAtta commented on src/main/java/games/strategy/engine/framework/GameRunner.java:
```diff
@@ -139,9 +125,12 @@ public static void main(final String[] args) {
-    SwingUtilities.invokeLater(LookAndFeel::setupLookAndFeel);
 
-    HttpProxy.setupProxies();
+    LookAndFeel.setupLookAndFeel();
```
This one was a bit interesting, `setupLookAndFeel` is an `invokeAndWait` call, so no point in wrapping it in an `invokeLater` call.

DanVanAtta replied:
I did notice two of these. I do believe it needs to be called before the Swing components are created.

The significant UI change is likely the UI theme default to the system default. That was updated intentionally to avoid special casing Mac OS behavior. I'm open to updating the default back to what it was, but note we had an OS specific setting there.

ssoloff replied:
> I do believe it needs to be called before the Swing components are created.

Right.  The unfortunate thing about this `main()` method is it's dealing with two orthogonal UI systems: Swing and JFX.  Regardless of where this call has to be to make the Swing client work correctly, my only concern is that it might have to be wrapped in a `!useJavaFxUi` guard.  Then we would have three such guards in this method. 😞

DanVanAtta replied:
Interesting one here. JFX is still a beta feature, it should also play well with this call as well (JFX was built to be able to integrate into existing Swing components). I think it is the better approach to leave the common path as unaltered as we can for the moment. Considering the case of many beta features, if we optimistically locked out all code that might not run, we would likely be over-complicating things. 

In this case specifically, we would remove `LookAndFeel.java` probably completely when JavaFx comes along, beyond sweeping for Swing code to remove, I'm pretty confident we will not forget to remove the call here.

DanVanAtta commented on src/main/java/games/strategy/engine/framework/headlessGameServer/HeadlessGameServer.java:
```diff
@@ -743,36 +744,10 @@ private static void handleHeadlessGameServerArgs() {
     }
-    // no passwords allowed for bots
-    // take any actions or commit to preferences
-    final String clientWait = System.getProperty(GameRunner.TRIPLEA_SERVER_START_GAME_SYNC_WAIT_TIME, "");
```
A lot of this was more complicated than need be. We already have a very slow timeout, and we adjust it by less than 5%. I removed most of this since the adjustments are not that significant. Most of it was outright removed as our `ClientSetting` infrastructure supports it pretty directly.

DanVanAtta commented on src/main/java/games/strategy/engine/framework/lookandfeel/LookAndFeel.java:
```diff
@@ -1,25 +1,74 @@
 
 public class LookAndFeel {
+
+  public static List<String> getLookAndFeelAvailableList() {
```
Some moved code, I thought it better belonged here, better cohesion.

DanVanAtta commented on src/main/java/games/strategy/engine/framework/lookandfeel/LookAndFeel.java:
```diff
@@ -38,39 +87,4 @@ public static void setupLookAndFeel() {
     });
   }
 
-  private static String getDefaultLookAndFeel() {
```
I removed this and now default all OS clients to ` UIManager.getSystemLookAndFeelClassName();` instead of ` SubstanceGraphiteLookAndFeel.class.getName();` for non-mac and  `UIManager.getSystemLookAndFeelClassName();`for mac.

I can restore that behavior if there is major objection, I suspect though we can just keep it simple. I can't think of a very strong reason for Mac OS to be special cased here.

DanVanAtta commented on src/main/java/games/strategy/engine/framework/startup/launcher/ServerLauncher.java:
```diff
@@ -248,13 +248,12 @@ public void run() {
           if (m_headless) {
             try {
               System.out.println("Game ended, going back to waiting.");
-              if (m_serverModel != null) {
```
`if (m_serverModel != null) {` is always true

DanVanAtta commented on src/main/java/games/strategy/engine/framework/startup/mc/ServerModel.java:
```diff
@@ -384,27 +384,15 @@ public void changeServerGameTo(final String gameName) {
-        save =
-            new File(ClientContext.folderSettings().getSaveGamePath(), SaveGameFileChooser.getAutoSaveEvenFileName());
-      } else {
+    public void changeToLatestAutosave(final SaveGameFileChooser.AutoSaveType autoSaveType) {
```
This was refactored after I noted that the only thing that changed was the second arg in the `File` constructor, for example ` SaveGameFileChooser.getAutoSaveFileName()`. Notably to refactor this those values/methods were moved to the `AutoSaveType` enum.

DanVanAtta commented on src/main/java/games/strategy/engine/framework/startup/ui/EnginePreferences.java:
```diff
@@ -1,246 +0,0 @@
-package games.strategy.engine.framework.startup.ui;
```
This class supported the engine preferences window that opened from the main screen.
The sound control was dropped for now, there is one available in game. Adding the same options to the settings window I think will make sense.

The look and feel option was moved to settings.

The set memory window button was broken!

IN the end all components were removed, we were left with launch map makers and console. I moved these to the main window and put them behind the beta feature flag. Both map makers need work, and for now it is an okay place to hide the console. Could use some suggestions for different or longer term treatments. I suspect this will be okay medium term while we evaluate the map makers and decide on any potential new home for the launch console button.

DanVanAtta commented on src/main/java/games/strategy/triplea/settings/HasDefaults.java:
```diff
@@ -1,5 +0,0 @@
-package games.strategy.triplea.settings;
```
This interface was used to support the old settings framework. It was a brittle to get working and was pretty locked in to the existing settings UI structure.  I was never very happy with it, so the framework for this was rewritten, now we have all such configuration in one or two enums. It's still not super simple, but simpler and way more flexible than the previous framework.

DanVanAtta commented:
In last couple commits, added an in-game override for the map listing file and lobby URL. No more monkeying around with config files to connect to the test lobby :confetti_ball: 

![testing_panel](https://user-images.githubusercontent.com/12397753/28651843-3313924c-7239-11e7-9562-338fe63ab0e1.png)

RoiEXLab commented on src/main/java/games/strategy/triplea/settings/ClientSetting.java:
```diff
@@ -0,0 +1,144 @@
+ * ClientSetting.flush();
+ * </pre></code>
+ */
+public enum ClientSetting {
```
I'm not sure if this even works, but this might make stuff easier:
If the enums stored a function object along with the default value which builds the corresponding object from a string value, `#value()` could always return the right object type of the setting, the same could be done reversed for storing values

DanVanAtta replied:
If the corresponding object refers to the UI pair, the proxy configuration threw things for loop. For a long while there was only one enum with all configuration. I then found we need more sophisticated logic in the UI component, not having that be a reference when initializing the property enum is very useful. This avoids forward and circular reference problems, but also I found there was potentially a 1:n relationship.

RoiEXLab commented:
Travis is failing due to compile errors in the test code.
It seems that a variable was not properly removed

DanVanAtta commented:
Yeah, this is still a bit rough, but pretty close, close enough for an early look.

ssoloff commented on src/main/java/games/strategy/engine/framework/startup/mc/IServerStartupRemote.java:
```diff
@@ -34,7 +34,7 @@
   void changeServerGameTo(final String gameName);
 
-  void changeToLatestAutosave(final SaveGameFileChooser.AUTOSAVE_TYPE typeOfAutosave);
+  void changeToLatestAutosave(final SaveGameFileChooser.AutoSaveType typeOfAutosave);
```
Just flagging this out of extreme caution because this is a remote interface...

I don't know how RMI serializes enums, but if the enum name is sent across the wire, this may break compatibility.  Might have to run a network game with an old version that triggers this code path to see if it's an issue.

ssoloff commented on src/main/java/games/strategy/engine/framework/startup/ui/MetaSetupPanel.java:
```diff
@@ -95,6 +100,24 @@ private void layoutComponents() {
+
+      final JButton m_mapXmlCreator = SwingComponents
+          .newJButton("[Beta] Run the Map Creator", e -> ProcessRunnerUtil.runClass(MapXmlCreator.class));
+      add(m_mapXmlCreator, new GridBagConstraints(0,  11, 1, 1, 0, 0, GridBagConstraints.CENTER, GridBagConstraints.NONE,
```
Checkstyle: "Line is longer than 120 characters (found 121)."

ssoloff commented on src/main/java/games/strategy/engine/framework/startup/ui/MetaSetupPanel.java:
```diff
@@ -95,6 +100,24 @@ private void layoutComponents() {
+          new Insets(10, 0, 0, 0), 0, 0));
+
+
+      final JButton console = SwingComponents.newJButton("Show Console", e -> ErrorConsole.getConsole().setVisible(true));
```
Checkstyle: "Line is longer than 120 characters (found 122)."

ssoloff commented on src/main/java/games/strategy/engine/framework/startup/ui/MetaSetupPanel.java:
```diff
@@ -95,6 +100,24 @@ private void layoutComponents() {
+      add(mapCreator, new GridBagConstraints(0, 10, 1, 1, 0, 0, GridBagConstraints.CENTER, GridBagConstraints.NONE,
+          new Insets(10, 0, 0, 0), 0, 0));
+
+      final JButton m_mapXmlCreator = SwingComponents
```
I know this was just copy-pasted from the engine preferences dialog, but might as well remove the `m_` prefix since this line was changed.

ssoloff commented on src/main/java/games/strategy/triplea/settings/SelectionComponent.java:
```diff
@@ -0,0 +1,39 @@
+
+/**
+ * A SelectionComponent represents a UI component that a user can use to update the value of a ClientSetting.
+ * Instances of this class are created in: @{code SelectionComponentFactory}
```
Typo: `@{code` -> `{@code`

ssoloff commented on src/main/java/games/strategy/triplea/settings/SettingsWindow.java:
```diff
@@ -1,195 +1,202 @@
-        new AiTab(ClientContext.aiSettings()),
-        new ScrollSettingsTab(ClientContext.scrollSettings()),
-        new FoldersTab(ClientContext.folderSettings())));
+enum SettingsWindow {
```
A few observations I noted when playing around with this UI:

1. Parent frame for the dialog is not set, so you get some undesirable UI artifacts like a) the default Java system icon, and b) a new item in the system tray.  Might have to add a `GameRunner#newJDialog()` method in order to capture the main frame like we've been doing recently.
1. The feature added in #1511 to display the setting default/range appears to no longer be present.  Was that intentional?  If so, should we not at least restore the legacy behavior to display the default/range, if applicable, in the description text area?

**EDIT:** Nevermind about (2).  I see that you changed the UI to display the valid range in a tooltip rather than directly in the panel.  The default is no longer displayed, but I guess that is less important.

ssoloff commented on src/main/java/games/strategy/ui/SwingComponents.java:
```diff
@@ -324,34 +323,38 @@ public static JMenu newJMenu(final String menuTitle, final KeyboardCode keyboard
+      fileChooser.setFileSelectionMode(JFileChooser.FILES_ONLY);
+    }
+
+    final int result = fileChooser.showOpenDialog(null);
```
Similar issue reported above about not having a proper parent frame.  There's already a few file chooser methods in `GameRunner`.  I don't like to keep throwing UI stuff over there, but that seems to be the short-term solution we've chosen to address this problem...

ssoloff commented on src/main/java/games/strategy/engine/config/client/GameEnginePropertyReader.java:
```diff
@@ -54,22 +55,38 @@ public GameEnginePropertyReader() {
-      final String backupAddress =
-          propertyFileReader.readProperty(GameEnginePropertyReader.PropertyKeys.LOBBY_BACKUP_HOST_ADDRESS);
-
-      return backupPropertyFetcher.parseBackupValuesFromEngineConfig(backupAddress);
```
The `backupPropertyFetcher` field in this class appears to no longer be used.

DanVanAtta commented:
Note default UI them is now system default for all OS. I've not strong objection to changing it back to picking a specific substance theme for linux + windows. The simplicity of non-OS picks is very nice.

I'm testing around with some of the substance look and feels and some of the themes seem broken. I've always come back and wondered about Substance, whether it is worth it. It is a good bit prettier in some cases... :diamond_shape_with_a_dot_inside:

RoiEXLab commented on src/main/java/games/strategy/triplea/settings/ClientSetting.java:
```diff
@@ -0,0 +1,144 @@
+
+
+  ClientSetting() {
+    this.defaultValue = "";
```
You should probably reuse other constructors here: `this("");`

RoiEXLab commented on src/main/java/games/strategy/triplea/settings/ClientSetting.java:
```diff
@@ -0,0 +1,144 @@
+  }
+
+  ClientSetting(final File file) {
+    this.defaultValue = file.getAbsolutePath();
```
You should probably reuse other constructors here: `this(file.getAbsolutePath());`

RoiEXLab commented:
@DanVanAtta the build is failing due to some checkstyle violations

ssoloff commented on src/test/java/games/strategy/engine/config/client/LobbyPropertyFileParserTest.java:
```diff
@@ -64,7 +57,7 @@ public void parseWithSimpleCase() throws Exception {
     final File testFile = createTempFile(testProps);
 
-    final LobbyServerProperties result = testObj.parse(testFile, new Version(TestData.clientCurrentVersion));
+    final LobbyServerProperties result = LobbyPropertyFileParser.parse(testFile, new Version(TestData.clientCurrentVersion));
```
Checkstyle: "Line is longer than 120 characters (found 125)."

ssoloff commented on src/test/java/games/strategy/engine/config/client/LobbyPropertyFileParserTest.java:
```diff
@@ -90,7 +83,7 @@ private File createTempFile(final TestProps... testProps) throws Exception {
     final File testFile = createTempFile(testDataSet());
 
-    final LobbyServerProperties result = testObj.parse(testFile, new Version(TestData.clientCurrentVersion));
+    final LobbyServerProperties result = LobbyPropertyFileParser.parse(testFile, new Version(TestData.clientCurrentVersion));
```
Checkstyle: "Line is longer than 120 characters (found 125)."

RoiEXLab commented:
@ssoloff Hmm the JavaFX Client looks fine to me, the only thing I noticed is that I apparently never commited a font file because of a gitignore setting I guess...
I will commit the font once I'm back home, because I currently have no idea how javafx loads fonts 😅

ssoloff commented:
@RoiEXLab Sorry, I should have specifically mentioned that it was the font that I noticed had changed.  When I run the JFX client, the system font is being used.

(I could have sworn I saw the font from your screenshots when I tested the JFX client PR on my machine.   Not sure how that would be possible if the font files were never committed. ¯\\\_(ツ)\_/¯)

RoiEXLab commented:
@ssoloff No Problem,
I actually figured out what the problem was, I opened #2142 to adress this issue.
It recall installing the font onto my system, in order for it to work with SceneBuilder, that's why it worked without the file in TripleA for me...

RoiEXLab commented:
@DanVanAtta checkstyle is still failing... There are too many violations in the test code

DanVanAtta commented on src/main/java/games/strategy/triplea/settings/ClientSetting.java:
```diff
@@ -0,0 +1,144 @@
+  }
+
+  public void save(final String newValue) {
+    Preferences.userNodeForPackage(ClientSetting.class).put(name(), newValue);
```
Took a shot at this, the helper method does not seem very clean/easy to understand. For me a reader has to notice the three helper methods repeated, and then also read the one-liner, in this case it feels easier to see the three one-liners. If the file were longer and things mroe scattered, I may feel otherwise. At some point it may turn out that the two `Preferences.userNodeForPackage(ClientSetting.class)` calls here might turn into the only two calls to `Preferences.userNodeForPackage`

ssoloff commented on src/main/java/swinglib/JButtonModal.java:
```diff
@@ -0,0 +1,70 @@
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+
+public class JButtonModal {
```
Should this be `JButtonModel`?

DanVanAtta replied:
yes, it is fixed locally, next push would have it fixed

ssoloff commented on src/main/java/swinglib/JButtonModal.java:
```diff
@@ -0,0 +1,70 @@
+  }
+
+
+  public static class JButtonModalBuilder {
```
Just thinking out loud...  Is there a need for both `JButtonModel` and `JButtonModelBuilder`?  It seems you only need the builder class, or at least `JButtonModel` could somehow be an implementation detail.

Also, do you perceive more factory methods in the future in addition to `swingComponent()`?  I didn't know if that's the reason you called it something other than `build()`.  The name `swingComponent()` wouldn't have been obvious to me.

If you can get rid of or hide the intermediate `JButtonModel` class, the API would be more like:

```java
final JButton button = JButtonBuilder.newInstance()
    .withTitle(...)
    .withTooltip(...)
    .withActionListener(...)
    .build();
```

I didn't test it, but the following compiles cleanly with 33% fewer lines (the `build()` method probably needs improvement):

```java
package swinglib;

import static com.google.common.base.Preconditions.checkNotNull;

import javax.swing.JButton;

import com.google.common.base.Preconditions;
import com.google.common.base.Strings;

public final class JButtonBuilder {
  private String title;
  private String toolTip;
  private Runnable actionListener;

  private JButtonBuilder() {}

  public static JButtonBuilder newInstance() {
    return new JButtonBuilder();
  }

  /** required - The text that will be on the button. */
  public JButtonBuilder withTitle(final String title) {
    Preconditions.checkArgument(!Strings.nullToEmpty(title).trim().isEmpty());
    this.title = title;
    return this;
  }

  /** optional, but potentially required in the future. This is the hover text when hovering on the button. */
  public JButtonBuilder withToolTip(final String toolTip) {
    Preconditions.checkArgument(!Strings.nullToEmpty(toolTip).trim().isEmpty());
    this.toolTip = toolTip;
    return this;
  }

  /** request, the event that occurs when the button is clicked */
  public JButtonBuilder withActionListener(final Runnable actionListener) {
    this.actionListener = checkNotNull(actionListener);
    return this;
  }

  public JButton build() {
    final JButton button = new JButton(Preconditions.checkNotNull(title));
    if (toolTip != null) {
      button.setToolTipText(toolTip);
    }
    button.addActionListener(e -> Preconditions.checkNotNull(actionListener).run());
    return button;
  }
}
```

DanVanAtta replied:
Great suggestion, thank you. I'll incorporate it in. It's pretty slick that the redundant variables were eliminated.

DanVanAtta commented:
@ssoloff this branch should be more stable now, please have a look when you can

Cernelius commented:
> Hid the 'launch map creator' buttons behind the beta feature flag. Those are both quite buggy, IMO we need some more work/time to decide what is useful and how to handle those applications.

?
As far as I know the non-beta one is not buggy at all and it seems close to perfect to me.

Cernelius commented:
If you can redirect me to the bug reports for the so called non-beta map creator I might see to double check them. I'm just saying I've used all of them and never had any bugs, but not very recently so.

DanVanAtta commented:
@Cernelius thanks, good to know. I was not clear on the final state of where the map creators were. It was pretty easy to unhide, that is now done:

![new_maikn](https://user-images.githubusercontent.com/12397753/28740757-d8237a2e-73bd-11e7-916e-72e1fd4f4efd.png)

If beta is turned on, then main looks like this:
![beta_flag](https://user-images.githubusercontent.com/12397753/28740761-e138e964-73bd-11e7-83ca-5887606fab76.png)

Cernelius commented:
Can be as part of this thing be done the names changes I proposed. None of those two is really a "map creator".
https://github.com/triplea-game/triplea/issues/1860
"Run the Map Creator"->"**Run the Main Map Making Utilities**"
"[Beta] Run the Map Creator"->"**[Alfa] Run the XML Creator**"
"TripleA Map XML Creator"->"**TripleA XML Creator**"

Cernelius commented:
In particular, calling them both "Map Creator" seems like the beta is the new and the other one is the old (at least I believe this is what about everybody would assume), while none of them is a map creator, and the new is just about making the whole xml (and nothing else but the xml), while the old "creator" is mostly not about the xml but, while not making the xml itself, it has some important xml related utilities (practically, it fully covers everything that it is really too time expensive to do it manually) that, then, you manually copy-paste inside your xml.
So, while having both the same name hints that they should be about the same things, they actually overlaps very little (the beta one is more about doing what the other one doesn't cover that much, namely the xml) and are conceptually very dissimilar.

DanVanAtta commented:
@Cernelius not quite at this junction, to respond to each proposed change:
1. Map Creator, I think there is an overhaul to be done there. For the moment I'd like to avoid further changes, the changes to it thus far were motived by killing off the prevoius 'engine preferences' screen
2. Shrug, we can discuss future name changes, first let's discuss what to do with the beta map creator and *all* that code! For now the button is now a true beta feature, visible only when the setting is turned on. It does not really matter what the name of the button is, similar for the map creator button.
3. "TripleA Map XML Creator"->"TripleA XML Creator"; Seems fine, though no reason to do it here. See https://github.com/triplea-game/triplea/pull/2151.

DanVanAtta commented:
>  (at least I believe this is what about everybody would assume),

The newest one is now behind the beta feature setting. Most people should not see it. The name for the button does not matter at this moment IMO. I'm very interested to know if there any components of it though that are actually useful. There is a lot of coding infrastructure for it, it would be great to get it back out of beta, and maybe strip it down to just the stuff that is useful (if anything)

DanVanAtta commented:
Ah, I should mention about:

> "Run the Map Creator"->"Run the Main Map Making Utilities"

With the other 'map creator' behind the beta feature flag, there would only appear to be one map creator to most users. So calling it 'the main' when it's not obvious where the other one is may even be confusing.

Cernelius commented:
I've never actually used the beta one and only superficially tested it, mainly because it appeared just broken to me. @FrostionAAA reviewed it deeply here:
https://github.com/triplea-game/triplea/files/1099902/BetaMapCreatorTesting.docx
but refused to actually use it for any of its projects.
Personally, I don't like the concept of an utility that makes the whole xml or modify it directly. I prefer to edit it myself or paste myself the stuff in it, and what it is not covered by the basic utilities it is well covered by using the in game "Export game.xml File" (for example, you make the whole setup, using edit mode, then export the xml), but this is my preference, and I guess that something for making the whole xml from start to finish may have a value at least as a tutor, if it works.

Cernelius commented:
Right. If the other utility is not available default, then:
"Run the Map Creator"->"**Run the Map Making Utilities**"
Probably that name started to be a bit too long anyways, and the "main" is not really needed. Mostly wanted to point out that doesn't allow you to do everything, but you need to do some stuff manually, but all the main things you need are fully covered.

Cernelius commented:
Are you sure you want to hide the "Show Console"? Doesn't matter to me or mapmakers than want to track stuff, as we will have the beta stuff on, but that is going to make bug reporting harder (but probably normal users will not think about clicking on there anyways).

Cernelius commented:
Also, off topic here but just to redirect you, regarding the "Show Console", here I've pointed out that I had to click on it to get the error report at all:
https://forums.triplea-game.org/topic/249/why-does-this-place-txt-file-not-work
@FrostionAAA got stuck because he didn't think about clicking on "Show Console", to know what the problem was, and that button is needed to get the info for bug reporting, if you closed the console.

DanVanAtta commented:
@Cernelius re: beta map creator, Thanks for posting the doc link, it seems like all the useful features in it are broken. Seems like we should stop supporting broken code.  I spun up a thread about it here: https://forums.triplea-game.org/topic/266/beta-map-creator-useful

> Are you sure you want to hide the "Show Console"? 

For now yes. It was buried behind another screen, which basically made it hidden. Needing to explicitly show the console during the staging screens is a bit exceptional. Hiding it for now I think is reasonable while we evaluate what we want to do with that feature in terms of making the console visible or not)

> Re: https://forums.triplea-game.org/topic/249/why-does-this-place-txt-file-not-work

That is arguably a coding error. We classify errors as usually one of two things:
1. Bad, but the things should still work, so quietly write down a message so someone watching would see it, otherwise presumably nobody should notice
2. As bad as it gets, console is then shown with error, with luck error handling is reasonable and the game can still go on even though something failed

ssoloff commented on src/test/java/swinglib/JButtonBuilderTest.java:
```diff
@@ -0,0 +1,70 @@
+  }
+
+  @Test(expected = IllegalArgumentException.class)
+  public void titleIsRequired() {
```
Quick question about these types of tests...  Is the intention to ensure `build()` throws an exception if the title is anything other than a non-empty string?  Or if `title()` should throw an exception under the same circumstances?

If the former, then this test might be passing for the wrong reason.  It appears that `title()` is throwing `IllegalArgumentException`, and the test never calls `build()`.  Would it be more correct to just not call `title()` in this case to ensure `build()` flags the unspecified required property?  (The code in `build()` looks like it only checks to see if the field is non-`null`, so it ends up throwing `NullPointerException` if the `title()` call is omitted.)

If the latter, then could the test be simplified to just `JButtonBuilder.builder().title("")`?

Maybe both cases need tests?

DanVanAtta replied:
There really should be two tests and not just the one. The intent of this test is that we get an error if we call `build()` without having called `title`.

The second test is exactly this one, ie: if you do specify a title, then it should be non-empty

I'll take a pass at looking at these.

ssoloff replied:
The latest change looks good, and the test names are very clear. :+1:

I still have a small concern about the body of `titleCannotBeEmpty()`.  If we only care about the behavior of `title()`, I find having to call all the additional setters required by `build()` to be a bit distracting.  In addition, the test could falsely pass if, for example, `title()` didn't throw but `actionListener()` did.

I could be missing some intent here, but is the following test equivalent?

```java
@Test(expected = IllegalArgumentException.class)
public void titleCannotBeEmpty() {
  JButtonBuilder.builder().title("");
}
```

If so, it would save some lines, and the next dev to come along and add a new builder property wouldn't think they had to go back and retrofit all the existing setter tests with calls to their newly-added setter.  IMO, only `build()` tests should require calls to setters for all required properties; the setter tests themselves should be focused on, well, themselves. :smile:

ssoloff commented on src/test/java/swinglib/JPanelBuilderTest.java:
```diff
@@ -0,0 +1,88 @@
+    assertThat(borderInsets.top, Is.is(borderWidth));
+    assertThat(borderInsets.bottom, Is.is(borderWidth));
+    assertThat(borderInsets.left, Is.is(borderWidth));
+    assertThat(borderInsets.right, Is.is(borderWidth));
```
Nit: These tests would read more easily if `org.hamcrest.core.Is.is` or `org.hamcrest.Matchers.is` were statically imported.

ssoloff commented on src/test/java/games/strategy/triplea/settings/SaveFunctionTest.java:
```diff
@@ -0,0 +1,114 @@
+
+import com.google.common.collect.ImmutableMap;
+
+@RunWith(MockitoJUnitRunner.class)
```
We've recently switched all Mockito test runners to `MockitoJUnitRunner.StrictStrubs.class`.

ssoloff commented on src/test/java/games/strategy/triplea/settings/ResetFunctionTest.java:
```diff
@@ -0,0 +1,38 @@
+import org.mockito.Mockito;
+import org.mockito.junit.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
```
We've recently switched all Mockito test runners to `MockitoJUnitRunner.StrictStrubs.class`.

ssoloff commented:
@DanVanAtta My recent comments are based on just browsing the commits you submitted over the past several hours.  I don't think I have time to do a full review of this PR today, but I'll do the review and some UX testing tomorrow.

DanVanAtta commented:
Thanks for the ongoing feedback, the latest looks good and I'll incorporate it in for simpler tests.

This is not necessarily an easy diff to review - thanks for tackling it!

DanVanAtta commented on game_engine.properties:
```diff
@@ -4,15 +4,3 @@
-
-
-## Location where we download latest lobby description file, this tells us where the latest lobby is located
-lobby_properties_file_url = https://raw.githubusercontent.com/triplea-game/triplea/master/lobby_server.yaml?raw=true
```
once we have the back removed, we are just left with two URLs. IT actually turns out we may as well hardcode them. They are now system settings, and we have a default value for them, if a user wants to change them, they can do so via the UI. No need for a  props file. If we want to release to update the values in this file, it is just as well the values be in code.

ssoloff replied:
> No need for a props file. If we want to release to update the values in this file, it is just as well the values be in code.

What about the JFX client and serialization proxy feature toggles?  If we want users to be able to try out those features, they're not going to be able to update code to enable them.  Am I misunderstanding the proposal here?

DanVanAtta replied:
Sorry I did not quite convey the full idea. There are no changes for anything other than two properties removed. We currently have the engine version, that is the only thing left in the game_engine.properties file.

For the feature toggles, IMO we should include them with on/off setting so it's easier to update. Knowing what the right property key is a small challenge.  Regardless, that capability is still there. It's just that the only thing that remains after this update in the `game_engine.properties` file after this update is version number.

For the two URLs, the backup URl was effectively a hack for us to override the lobby URL. Hence, for those two fields, after removing how the backup URL was done and allowing the URLs to be updated in settings UI, the initial value may as well come from code (hence the props removal).

IMO the JFX and proxy feature toggles would be well moved to the new settings UI and mimic the beta feature flag. By adding a couple enums you would get a radio button that could turn those features on/off. game_engine.properties is though not a bad answer for props that need to be enabled on startup.

ssoloff replied:
Thanks for the clarification.

> game_engine.properties is though not a bad answer for props that need to be enabled on startup.

The JFX client definitely falls into this category.

I think the serialization proxy stuff, at least for save games, can be moved, as you suggested, to the beta settings.  That's actually an improvement from the users' perspective, IMO.  I'll look into doing that once this PR is merged.

ssoloff commented on src/main/java/games/strategy/triplea/ui/TabbedProductionPanel.java:
```diff
@@ -27,6 +27,7 @@
 import games.strategy.ui.SwingComponents;
```
Static analysis: "The import games.strategy.ui.SwingComponents is never used"

ssoloff commented on src/main/java/games/strategy/ui/SwingComponents.java:
```diff
@@ -59,8 +56,111 @@
+    };
+  }
+
+  public enum KeyDownMask {
```
Checkstyle: "Missing a Javadoc comment."

ssoloff commented on src/main/java/games/strategy/ui/SwingComponents.java:
```diff
@@ -59,8 +56,111 @@
+    addKeyListener(component, KeyStroke.getKeyStroke(KeyEvent.VK_ENTER, 0), runnable);
+  }
+
+  public static void addEscapeKeyListener(
```
Checkstyle: "Missing a Javadoc comment."

ssoloff commented on src/main/java/games/strategy/ui/SwingComponents.java:
```diff
@@ -59,8 +56,111 @@
+  public static void addEscapeKeyListener(
+      final RootPaneContainer dialog,
+      final Runnable keyDownAction) {
+    if(dialog.getRootPane() != null) {
```
Checkstyle: "WhitespaceAround: 'if' is not followed by whitespace. Empty blocks may only be represented as {} when not part of a multi-block statement (4.1.3)"

ssoloff commented on src/main/java/games/strategy/ui/SwingComponents.java:
```diff
@@ -59,8 +56,111 @@
+      addKeyListener(dialog.getRootPane(), KeyStroke.getKeyStroke(KeyEvent.VK_ESCAPE, 0), keyDownAction);
+    }
+  }
+  public static void addEscapeKeyListener(
```
Checkstyle:

* "'METHOD_DEF' should be separated from previous statement."
* "Missing a Javadoc comment."

ssoloff commented on src/main/java/games/strategy/ui/SwingComponents.java:
```diff
@@ -59,8 +56,111 @@
+      addKeyListener(component.getRootPane(), KeyStroke.getKeyStroke(KeyEvent.VK_ESCAPE, 0), keyDownAction);
+    }
+  }
+  public static void addKeyListenerWithMetaAndCtrlMasks(
```
Checkstyle: "'METHOD_DEF' should be separated from previous statement."

ssoloff commented on src/main/java/games/strategy/ui/SwingComponents.java:
```diff
@@ -305,53 +343,45 @@ public static void showDialog(final String title, final String message) {
     return menu;
   }
 
+  public enum FolderSelectionMode {
```
Checkstyle: "Missing a Javadoc comment."

ssoloff commented on src/main/java/swinglib/GridBagHelper.java:
```diff
@@ -0,0 +1,78 @@
+
+  /**
+   * Adds many components in one go, a convenience api {@see add}
+   */
```
Checkstyle: "First sentence of Javadoc is incomplete (period is missing) or not present."

ssoloff commented on src/main/java/swinglib/JButtonBuilder.java:
```diff
@@ -0,0 +1,67 @@
+    return this;
+  }
+
+  /** request, the event that occurs when the button is clicked */
```
Checkstyle: "First sentence of Javadoc is incomplete (period is missing) or not present."

ssoloff commented on src/main/java/swinglib/JDialogBuilder.java:
```diff
@@ -0,0 +1,77 @@
+    return this;
+  }
+
+  public JDialogBuilder title(final String title) {
```
Checkstyle: "Missing a Javadoc comment."

ssoloff commented on src/main/java/swinglib/JPanelBuilder.java:
```diff
@@ -0,0 +1,191 @@
+  private int gridColumns;
+  private BorderType borderType;
+  private int borderWidth;
+  private Float xAlignment;
```
Checkstyle: "Member name 'xAlignment' must match pattern '^[a-z][a-z0-9][a-zA-Z0-9]*$'."

DanVanAtta replied:
Hmm, I think we may want to update that rule, `xAlignment` does seem okay, but I did note the checkstyle violation.

ssoloff commented on src/main/java/swinglib/JPanelBuilder.java:
```diff
@@ -0,0 +1,191 @@
+          break;
+      }
+    }
+    switch (layout) {
```
Checkstyle: "switch without "default" clause."

ssoloff commented on src/main/java/swinglib/JPanelBuilder.java:
```diff
@@ -0,0 +1,191 @@
+        }
+      }
+    }
+    if(xAlignment != null) {
```
Checkstyle: "WhitespaceAround: 'if' is not followed by whitespace. Empty blocks may only be represented as {} when not part of a multi-block statement (4.1.3)"

ssoloff commented on src/main/java/swinglib/JPanelBuilder.java:
```diff
@@ -0,0 +1,191 @@
+    return this;
+  }
+
+  public JPanelBuilder xAlignmentCenter() {
```
Checkstyle: "Method name 'xAlignmentCenter' must match pattern '^[a-z][a-z0-9][a-zA-Z0-9_]*$'."

ssoloff commented on src/main/java/swinglib/JPanelBuilder.java:
```diff
@@ -0,0 +1,191 @@
+    DEFAULT, GRID, GRID_BAG, BOX_LAYOUT_HORIZONTAL, BOX_LAYOUT_VERTICAL
+  }
+
+  public enum BorderLayoutPosition {
```
Checkstyle: "Missing a Javadoc comment."

ssoloff commented on src/main/java/swinglib/JPanelBuilder.java:
```diff
@@ -0,0 +1,191 @@
+    DEFAULT, CENTER, SOUTH, NORTH, WEST, EAST
+  }
+
+  public enum BorderType {
```
Checkstyle: "Missing a Javadoc comment."

ssoloff commented on src/test/java/swinglib/JPanelBuilderTest.java:
```diff
@@ -0,0 +1,88 @@
+  }
+
+  @Test
+  public void xAlignmentCenter() {
```
Checkstyle: "Method name 'xAlignmentCenter' must match pattern '^[a-z][a-z0-9][a-zA-Z0-9_]*$'."

ssoloff commented:
@DanVanAtta I played around with the new settings UI.  Overall, I think the new settings UI and related UI changes work well and are intuitive.  Below are some minor issues I discovered while testing, as well as some comments for possible improvements.

<hr>

#### Lack of parent frame

I mentioned this in a review from a few days ago, but I just want to reiterate that using a `null` parent frame for `SettingsWindow` produces some annoying UI  issues:

* The settings window is initially placed in the top-left corner of the desktop instead of being centered over the main window.
* The settings window uses the default Java icon instead of the TripleA icon.
* An extra (and unnecessary) item appears in the system tray for the settings window.

#### Margins

Some margins around the various controls would go a long way to improving the overall L&F.  However, I think it's acceptable to do that in a follow-up PR.

#### Description field word wrap

The description field for each setting does not appear to wrap text (unless there happens to be an embedded newline).  For example, the descriptions on the Testing tab are particularly long and extend approximately an additional 33% off the right side of the window.  I have to scroll the entire panel to the right to read some of them:

![settings-text-field-width](https://user-images.githubusercontent.com/4826349/28749820-280ad4c6-74a2-11e7-9018-6fb55fa0d487.png)

#### Default setting value

I previously stated that no longer displaying the default value for each setting was probably not a big deal.  However, after using the UI, I'm no longer sure about that statement.

If I only want to change one setting back to its default, I used to be able to just change it manually to match the default value displayed on the screen.  Now, if I don't happen to know what the default value is, I have to reset all settings on the tab back to their default, and then change the other values that I didn't want to change back to their original values.  Displaying the default in the tooltip somehow could avoid this issue.

Unless, of course, I missed something obvious, and there's a perfectly reasonable way to easily reset one setting back to its default value. :smile:

#### Unsaved values not reset after closing settings window

1. Change a setting value.
1. Close settings window.
1. Re-open settings window.
1. Modified value is still present in the field, but it has not actually been saved.

For example, if I change the **Testing -> Lobby Host Override** setting to "localhost", close the settings window, and attempt to Play Online, I'll still be connected to the default lobby.  Even though, when I re-open the settings window, "localhost" is still displayed in the field.

If I exit TripleA, restart TripleA, and open the settings window, the original setting value is correctly displayed in the field.

#### Value constraint violation UI not updated after resetting to default value

1. Change a setting value so it violates one of its constraints (e.g. change **Network Timeouts -> Start game timeout** to 5000).
1. Remove focus from field and observe that background color is changed to red.
1. Click **Restore Defaults**.
1. Dismiss confirmation message.
1. Observe that field now contains default value but that background color is still red.

Giving focus to the field in question and then removing it restores the default background color.

#### Beta features not added/removed when returning to main menu

1. Change the **Testing -> Show Beta Features** setting.
1. Click **Save**.
1. Dismiss confirmation message.
1. Close settings window.
1. Observe that main menu has not been updated to add/remove the beta features.

If I go to a different screen (e.g. Start Local Game), and then return to the main menu via Cancel, the beta features are added/removed.

RoiEXLab commented:
@DanVanAtta some merge conflicts showed up.

DanVanAtta commented:
> Lack of parent frame

I took a pass at adding in parent frame references. In one or two cases I don't think it was possible without doing serious wiring work to pass frame references around.

The listing of issues with no parent frame is useful, hard to disagree with them, though there is one I think I may note:

> An extra (and unnecessary) item appears in the system tray for the settings window.
Margins

The extra item in the system tray is generally useful for me. If the extra window loses focus, clicking the main 'TripleA' icon in the system tray (on Ubuntu 16) does not bring all windows into focus. With many open windows, I generally have to systematically minimize windows to find the extra setting window.


I don't think we'll necessarily solve this here, and let me know if there are some remaining low hanging fruit I missed where we could add parent references. I think what we need is to create a really nice `JFrameBuilder` that can take care of these concerns without passing around JFrame references everywhere.


> Some margins around the various controls would go a long way to improving the overall L&F. However, I think it's acceptable to do that in a follow-up PR.

Agree on both points, it'll need some tweaking.



> Default setting value

I moved the 'reset' button to have one button next to each input component that resets only that component. The effect of that button is not persisted immediately, so 'save' is now the only thing tha persists values. Let me know what you think and if this addresses the UX concern.


> Unsaved values not reset after closing settings window

Good thing to catch, thank you for good repro steps. I fixed this just now. This was because we were creating swing components on enum instantiation, we then never recreated the components. I resovled this recently by converting the definitions to suppliers, renaming the `getComponent` method to be a `build` method (paraphrasing the exact names), and then added a `dispose` method to UIBindings to clean up when the window is closed. This then allowed for components to be recreated and to load the current value. 


> Value constraint violation UI not updated after resetting to default value

Same problem where state was not being reset, should also now be fixed.


> Beta features not added/removed when returning to main menu

I did not quite intend for the flag to function that way. Overall, an immediate effect would be nice, and I can see value in seeing different UI treatments. In practice, I think we are a long way from benefitting from that. Meanwhile, beta features I think would become a lot more complicated. Notably not only would a feature need to check to see if it should be enabled, but we would have to add a 'refresh' hook into anything that has a beta flag. The interest of minimal, simpler infrastructure takes precedence here IMO. 


------------

Most things should be addressed at this point. Some items that are known to remain, for a fast follow:
- UI padding
- Text description handling. I could use some help here, but the text description is likely not wrapping very nicely and breaking at the right spots
- Improved text

ssoloff commented:
> > Default setting value
>
> I moved the 'reset' button to have one button next to each input component that resets only that component. The effect of that button is not persisted immediately, so 'save' is now the only thing tha persists values. Let me know what you think and if this addresses the UX concern.

I like the new solution.  However, what you described is not the behavior I'm seeing.  The default value is immediately persisted when I click **Reset to Default** (no confirmation message is displayed).  I don't need to click **Save**.  It looks like the `SelectionComponent#resetToDefault()` implementations call `ClientSetting#flush()` after resetting the value.

> > Unsaved values not reset after closing settings window
>
> I fixed this just now.

Confirmed.  Thanks.

> > Value constraint violation UI not updated after resetting to default value
>
> Same problem where state was not being reset, should also now be fixed.

Hmmm.  This still seems to be happening for me.  I still have to trigger a lost focus event on the text field for the background color to be reset back to normal.

> > Beta features not added/removed when returning to main menu
>
> The interest of minimal, simpler infrastructure takes precedence here IMO.

Agreed.  I think the addition of the text informing the user they need to restart is sufficient for now.

ssoloff commented on src/main/java/games/strategy/triplea/settings/SettingsWindow.java:
```diff
@@ -1,195 +1,150 @@
 
-  private SettingsWindow(final SettingsTab<?>... tabs) {
-    add(buildTabbedPane(tabs), BorderLayout.CENTER);
+  public synchronized void open() {
```
For both `open()` and `close()`...  Is `synchronized` required here?  These methods should always be called on the EDT.  There are deeper problems if they are called off the EDT because the Swing components could potentially be corrupted.  Since these methods are API, you might want to consider calling `Preconditions.checkState(SwingUtilities.isEventDispatchThread())` at the top of these methods to fail fast if the caller invokes them from a non-EDT thread.

DanVanAtta replied:
We do have a shared state problem on the `dialog` variable. `synchronized` is a bit defensive, but namely as you suggested if always on the EDT thread, then not really a problem. Thank you for the good suggestion of a `Preconditons` check.

ssoloff commented on src/main/java/games/strategy/triplea/settings/ClientSettingUiBinding.java:
```diff
@@ -0,0 +1,213 @@
+    return current().getJComponent();
+  }
+
+  private synchronized SelectionComponent current() {
```
Similar to my previous comment about the need for `synchronized` on `SettingsWindow#open()`...  This code isn't quite the same because `ClientSettingUiBinding` is not a Swing component.  However, the `Ui` in its name suggests to me that it is only called by UI code, and thus is effectively single threaded.  Is that not true?

If not, and the synchronization is required, then you might also want to synchronize `dispose()`.  While assignment to an object reference is atomic, the publication of the updated value to other threads is not guaranteed without some kind of memory barrier (unless the field happens to be `volatile`).

DanVanAtta replied:
Thanks, I remove synchronized. 56bf560

ssoloff commented on src/main/java/games/strategy/triplea/settings/ClientSettingUiBinding.java:
```diff
@@ -0,0 +1,213 @@
+  ClientSettingUiBinding(
+      final String title,
+      final SettingType type,
+      final Supplier<SelectionComponent> selectionComponent,
```
Should this parameter be named `selectionComponentBuilder`?  And should it have `checkNotNull()` applied?

DanVanAtta replied:
86a537f updates per suggestion. Was thinking a no-op runnable is effectively a null, we are not checking for that, so why also check for null? But, may as well just to keep symmetry.

ssoloff commented on src/main/java/games/strategy/engine/framework/startup/ui/MetaSetupPanel.java:
```diff
@@ -9,14 +9,24 @@
+/**
+ * This is the main welcome panel with 'play online' buttons etc...
+ * The panel does not include the 'play
+ */
```
Looks like you may have been interrupted while editing this comment (see 1e8e282). :smile:

ssoloff commented on src/main/java/games/strategy/triplea/settings/GameSettingUiBinding.java:
```diff
@@ -0,0 +1,45 @@
+  /**
+   * Reads the value set in the UI and return true if it is valid, false otherwise.
+   */
+  boolean isValid();
```
Checkstyle: "'METHOD_DEF' should be separated from previous statement."

ssoloff commented on src/main/java/swinglib/JDialogBuilder.java:
```diff
@@ -0,0 +1,85 @@
+   *
+   * @param title
+   * @return
+   */
```
This Javadoc is reporting the following Checkstyle warnings:

* Line 70: "First sentence of Javadoc is incomplete (period is missing) or not present."
* Line 72: "At-clause should have a non-empty description."
* Line 73: "At-clause should have a non-empty description."

ssoloff commented on src/main/java/games/strategy/triplea/settings/SelectionComponentFactory.java:
```diff
@@ -0,0 +1,407 @@
+    return selectFile(clientSetting, SwingComponents.FolderSelectionMode.DIRECTORIES);
+  }
+
+  static Supplier<SelectionComponent> selectionBox(final ClientSetting clientSetting, final List<String> availableOptions) {
```
Checkstyle: "Line is longer than 120 characters (found 124)."

ssoloff commented on src/main/java/games/strategy/triplea/settings/SaveFunction.java:
```diff
@@ -0,0 +1,75 @@
+  /**
+   * Returns a result message after persisting settings.
+   */
+  static SaveResult saveSettings(final Iterable<? extends GameSettingUiBinding> settings, final Runnable settingsFlushToDisk) {
```
Checkstyle: "Line is longer than 120 characters (found 127)."

ssoloff commented on src/main/java/games/strategy/triplea/settings/SettingsWindow.java:
```diff
@@ -1,195 +1,150 @@
-import games.strategy.triplea.settings.scrolling.ScrollSettingsTab;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
```
Static analysis: "The import com.google.common.base.Strings is never used"

ssoloff commented on src/main/java/games/strategy/engine/framework/map/download/MapDownloadProgressPanel.java:
```diff
@@ -15,6 +15,7 @@
 
 import games.strategy.engine.ClientContext;
 import games.strategy.ui.SwingComponents;
```
Bump...

ssoloff commented on src/main/java/games/strategy/triplea/ui/PlayersPanel.java:
```diff
@@ -9,14 +9,17 @@
 import games.strategy.engine.data.PlayerID;
 import games.strategy.engine.framework.IGame;
 import games.strategy.ui.SwingComponents;
```
Static analysis: "The import games.strategy.ui.SwingComponents is never used"

ssoloff commented on src/main/java/games/strategy/ui/SwingComponents.java:
```diff
@@ -1,53 +1,50 @@
 import javax.swing.SwingWorker;
-import javax.swing.UIManager;
 import javax.swing.border.Border;
 import javax.swing.border.EmptyBorder;
```
Static analysis:

* "The import javax.swing.border.Border is never used"
* "The import javax.swing.border.EmptyBorder is never used"

ssoloff commented on src/main/java/games/strategy/triplea/settings/ClientSetting.java:
```diff
@@ -0,0 +1,136 @@
+import java.util.prefs.BackingStoreException;
+import java.util.prefs.Preferences;
+
+import javax.swing.JFrame;
```
Static analysis: "The import javax.swing.JFrame is never used"

DanVanAtta commented:
@ssoloff Thanks again for the useful feedback. I think I've got everything raised so far now addressed. `resetToDefault` was a bit buggy, I broke it up to a `reset` and `reset to default` button, and put those on the south border of the settings window. With that I think all the UI behavior is fixed up.

On another note, the 'save' button probably should only be enabled when an option has been updated. (Though, I'd prefer to leave that for follow up branch).

ssoloff commented:
@DanVanAtta I saw that you just merged `master` into this branch.  I'm not sure if that's because you're preparing to add further commits.  I'll wait for a few hours before merging this PR just in case.  Feel free to self-merge if you're all done.

DanVanAtta commented:
Thanks, merging now.

>  I saw that you just merged master into this branch

The intent was simply to update to the latest.

ron-murhammer commented:
@DanVanAtta @ssoloff @RoiEXLab I think we need to consider setting a better default Look and Feel. I think the OS default is significantly worse than the old L&F we had. I don't care that much about the main menu and setup stuff but in game the right panel looks a lot worse. I think something darker would probably be preferred. Might be good to open this up to forum users as well.

FrostionAAA commented:
I have not been reading this long thread. But I have seen this mentioned on the forum. Would it not be possible to have this new UI with some options about colour. Like during installation the user can select colour?

DanVanAtta commented:
@FrostionAAA this is update you're looking for which would remove the substance look and feels. It's not yet live and still pending: https://github.com/triplea-game/triplea/pull/2165
